src/main/kotlin/cacophony/semantic/analysis/VariablesMap.kt

```diff
@@ -4,14 +4,199 @@ import cacophony.controlflow.Variable
 import cacophony.semantic.names.ResolvedVariables
 import cacophony.semantic.syntaxtree.AST
 import cacophony.semantic.syntaxtree.Assignable
+import cacophony.semantic.syntaxtree.Block
 import cacophony.semantic.syntaxtree.Definition
+import cacophony.semantic.syntaxtree.Empty
+import cacophony.semantic.syntaxtree.Expression
+import cacophony.semantic.syntaxtree.FieldRef
+import cacophony.semantic.syntaxtree.FunctionCall
+import cacophony.semantic.syntaxtree.Literal
+import cacophony.semantic.syntaxtree.OperatorBinary
+import cacophony.semantic.syntaxtree.OperatorUnary
+import cacophony.semantic.syntaxtree.Statement
+import cacophony.semantic.syntaxtree.Struct
+import cacophony.semantic.syntaxtree.StructField
+import cacophony.semantic.syntaxtree.VariableUse
+import cacophony.semantic.types.StructType
 import cacophony.semantic.types.TypeCheckingResult
+import cacophony.semantic.types.TypeExpr
 
 data class VariablesMap(
     val lvalues: Map<Assignable, Variable>,
     val definitions: Map<Definition, Variable>,
 )
 
 fun createVariablesMap(ast: AST, resolvedVariables: ResolvedVariables, types: TypeCheckingResult): VariablesMap {
-    TODO()
+    val definitions = VariableDefinitionMapBuilder(types).build(ast)
+    val lvalues = AssignableMapBuilder(resolvedVariables, definitions).build(ast)
+    return VariablesMap(lvalues, definitions)
+}
+
+private class AssignableMapBuilder(val resolvedVariables: ResolvedVariables, val definitions: Map<Definition, Variable>) {
+    private val lvalues = mutableMapOf<Assignable, Variable>()
+
+    fun build(ast: AST): Map<Assignable, Variable> {
+        visit(ast)
+        return lvalues
+    }
+
+    fun visit(expression: Expression) {
+        when (expression) {
+            is FieldRef.LValue -> visitFieldRefLValue(expression)
+
+            is VariableUse -> visitVariableUse(expression)
+
+            is Block -> expression.expressions.forEach { visit(it) }
+            is Definition.FunctionArgument -> {
+                /* do nothing */
+            }
+
+            is Definition.ForeignFunctionDeclaration -> {
+                /* do nothing */
+            }
+
+            is Definition.FunctionDefinition -> visit(expression.body)
+            is Definition.VariableDeclaration -> visit(expression.value)
+            is Empty -> {
+                /* do nothing */
+            }
+
+            is FieldRef.RValue -> {
```

Comment:
I think it fails on something like `(s).x`, as block is not lvalue itself, but can have lvalue inside, which we left unvisited
We should visit expression.obj here as well

Reply:
good catch, fixed